### PR TITLE
WIP fixes #13406 adds dhcp proxy tab and subnet details

### PR DESCRIPTION
**Pull request by ohadlevy**
additionally, this introduce a few refactors to the smart proxies controller

### CODE REVIEW

**Comment by ohadlevy** [comment]
look & feel example

recorded

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+</ul>
+<div id="proxy-dhcp-tab-content" class="tab-content">
+  <div class="tab-pane active in" id="dhcp_general">
+    <div class="row">
+      <%= show_feature_version('dhcp') %>
+    </div>
+
+    <div class="container-fluid container-cards-pf">
+      <div class="row row-cards-pf">
+        <% subnets.in_groups_of(2).each do |subnets| %>
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb]: This is failing badly when you have just one or an odd number of subnets - 
screenshot from 2016-01-27 12-13-20

File: app/views/smart_proxies/plugins/_dhcp_reservations.html.erb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    <tr>
+      <th class="col-md-5"><%= _("Hostname") %></th>
+      <th class="col-md-1"><%= _("IP Address") %></th>
+      <th class="col-md-2"><%= _("MAC") %></th>
+    </tr>
+  </thead>
+  <tbody>
+    <% reservations.each do |record| %>
+      <tr>
+        <td class="ellipsis"><%= record['hostname'] %></td>
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_reservations.html.erb]: This should link to the Foreman host whenever possible

File: app/views/smart_proxies/plugins/_dhcp_leases.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      <th class="col-md-2"><%= _("MAC") %></th>
+      <th class="col-md-2"><%= _("State") %></th>
+      <th class="col-md-2"><%= _("Start") %></th>
+      <th class="col-md-2"><%= _("End") %></th>
+    </tr>
+  </thead>
+  <tbody>
+    <% leases.each do |record| %>
+      <tr>
+        <td class="ellipsis"><%= record['hostname'] %></td>
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_leases.html.erb]: This should link to the Foreman host whenever possible

File: app/services/proxy_status/dhcp.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require 'ostruct'
```
Comment (dLobatog) [inline app/services/proxy_status/dhcp.rb]: What are you using ostruct for here?

File: app/helpers/smart_proxies_helper.rb [hunk 5]
```diff
@@ -65,4 +65,13 @@ def tabbed_features(proxy)
   def show_feature_version(feature)
     render :partial => 'smart_proxies/plugins/plugin_version', :locals => { :feature => feature }
   end
+
+  def subnet_label subnet
```
Comment (dLobatog) [inline app/helpers/smart_proxies_helper.rb]: Parenthesis missing to wrap the method arguments (thought rubocop would enforce this one)

**Comment by shlomizadok** [comment]
To avoid long horizontal tabs (when having lots of subnets), we can switch to vertical tabs:
image

image

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          <% subnets.each do |subnet| %>
+            <div class="col-md-4">
+              <div class="card-pf card-pf-aggregate-status card-pf-accented">
+                <h2 class="card-pf-title">
+                  <a href="#<%= subnet.network.parameterize %>">
+                    <%= subnet_label(subnet) %>
+                    <span class="card-pf-aggregate-status-count">
+                      <%= Host.joins(:interfaces => :subnet).where(:subnets => {:dhcp_id => @smart_proxy.id, :network => subnet.network }).count %>
+                    </span>
+                    <%= _('Hosts') %>
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb]: I think all figures, like # of hosts should go below as `aggregate-status-notifications`

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                <div class="card-pf-body">
+                  <p class="card-pf-aggregate-status-notifications">
+                    <a href="#<%= subnet.network.parameterize %>_reservation">
+                      <span class="fa fa-database" data-toggle="tooltip" title="<%= _('View Reservations') %>">
+                        <%= subnet.reservations.count %>
+                        <%= _('Reservations') %>
+                      </span>
+                    </a>
+                    <a href="#<%= subnet.network.parameterize %>_leases">
+                      <span class="fa fa-laptop" data-toggle="tooltip" title="<%= _('View Leases') %>">
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb]: The text shouldn't go in the `span` I think. This is causing the number of leases and reservations to show up with Font awesome letters, not Patternfly. See:

screenshot from 2016-01-27 12-25-48
screenshot from 2016-01-27 12-25-33

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                  <a href="#<%= subnet.network.parameterize %>">
+                    <%= subnet_label(subnet) %>
+                    <span class="card-pf-aggregate-status-count">
+                      <%= Host.joins(:interfaces => :subnet).where(:subnets => {:dhcp_id => @smart_proxy.id, :network => subnet.network }).count %>
+                    </span>
+                    <%= _('Hosts') %>
+                  </a>
+                </h2>
+                <div class="card-pf-body">
+                  <p class="card-pf-aggregate-status-notifications">
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb:39]: This card feels too bloated IMO, too much information. How about 1 card for leases, another for reservations, and another for hosts, and group them in one row per subnet?

One thing that could be interesting (not advocating for it to be added now) is list the number of IPs remaining, https://rawgit.com/patternfly/patternfly/master/tests/cards.html

File: app/controllers/smart_proxies_controller.rb [hunk 9]
```diff
@@ -93,16 +100,22 @@ def find_status
 
   def requested_data
     data = yield
-    render :json => {:success => true, :message => data }
+    render :json => { :success => true, :message => data }
+  rescue Foreman::Exception => exception
+    render :json => { :success => false, :message => exception.message } and return
+  end
+
+  def requested_data_html
```
Comment (dLobatog) [inline app/controllers/smart_proxies_controller.rb]: Request ajax data perhaps?

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb [hunk 10]
```diff
@@ -0,0 +1,69 @@
+<ul id="proxy-dhcp-tab" class="nav nav-tabs nav-tabs-pf">
+  <li class="active"><a href="#dhcp_general"><%= _('General') %></a></li>
+  <% subnets.each do |subnet| %>
+    <li><a href="#<%= subnet.network.parameterize %>"><%= subnet_label subnet %></a></li>
+  <% end %>
+</ul>
+<div id="proxy-dhcp-tab-content" class="tab-content">
+  <div class="tab-pane active in" id="dhcp_general">
+    <div class="row">
+      <%= show_feature_version('dhcp') %>
```
Comment (dLobatog) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb:10]: For some reason I never get the version here, it always shows a spinner

Comment (shlomizadok) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb:10]: Yeah, this happens because `showProxies()` is not called when this page loads (run it manually from console and it will work...)

(this is probably because we create the html from `data-ajax-url`)

Comment (ohadlevy) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb:10]: fixed by triggering 'ContentLoad' in ajax-url function

File: app/assets/javascripts/application.js [hunk 11]
```diff
@@ -123,6 +123,7 @@ function onContentLoad(){
       if ($(this).data('on-complete')){
         window[$(this).data('on-complete')].call(null, this, status);
       }
+      $(document).trigger('ContentLoad');
```
Comment (ohadlevy) [inline app/assets/javascripts/application.js]: @tbrisker any concerns about this?

Comment (tbrisker) [inline app/assets/javascripts/application.js]: Why is this needed? it will cause the `onContentLoad` function to run again for every ajax-url. If you need to handle some initialization (i.e. datatables) once the content loads, it would be better to add that in the data-on-complete attribute which is called once the loading completes.

Comment (ohadlevy) [inline app/assets/javascripts/application.js]: I need multiple hooks (setVersion, data table etc) imho that was the whole
idea of on content load (as new content is loaded) ?

Comment (tbrisker) [inline app/assets/javascripts/application.js]: You could use `data-on-complete="onContentLoad"`, or even better create a function that contains the hooks you need. 
I'm afraid adding this here may cause problems on other pages. This was the original idea of contentLoad, however over time it has become quite bloated and might cause breakage if executed twice on some pages (for example, IIRC a `select2` that has been initialized breaks if you try to initialize it again).

Comment (ohadlevy) [inline app/assets/javascripts/application.js]: :(

Comment (lzap) [inline app/assets/javascripts/application.js]: Take a look how I do it in Logs patch, I gave up on the attribute and wrote my own initialization, because my table gets totally different configuration. There is nothing commin, except the positioning of Filter drop down...

File: app/views/smart_proxies/plugins/_dhcp_subnets.html.erb (hunk 7 above)
Comment (lzap) [inline app/views/smart_proxies/plugins/_dhcp_subnets.html.erb]: Not a UX expert :-) but I think there is too much info in the card. Consider breaking those up, those should fit in 6 columns (grid). Maybe remove the "Reservations" and "Leases" text and put them into hover tip, icon should be enough.

**Comment by xprazak2** [comment]
@dLobatog: I have been trying to move this along a bit, can you take a look when you have a spare moment? [This](https://github.com/ohadlevy/foreman/pull/6) aims to address the comments above.

File: app/assets/javascripts/proxy_status/dhcp.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+function activateDhcpTables(state) {
+    showProxies();
+    activateDatatables();
```
Comment (ohadlevy) [inline app/assets/javascripts/proxy_status/dhcp.js:3]: maybe we should include setTabs(), and if we do, maybe we need a general status callback to trigger all of the above (as its seems common to most)?

File: app/helpers/smart_proxies_helper.rb [hunk 13]
```diff
@@ -65,4 +65,25 @@ def tabbed_features(proxy)
...
+  def link_to_host(subnet, hostname, smart_proxy)
+    foreman_subnet = Subnet.where(:network => subnet.network, :netmask => subnet.netmask)
+    return hostname unless foreman_subnet
+    host = hosts_with_subnet(subnet, smart_proxy, :name => hostname).first
+    host.present? ? link_to(host.to_s, host_path(host)) : hostname
+  end
+
+  def hosts_with_subnet(subnet, smart_proxy, condition = {})
+    subnets =  { :subnets => { :dhcp_id => smart_proxy.id, :network => subnet.network } }
+    Host.joins(:interfaces => :subnet).where(subnets.merge! condition)
```
Comment (ohadlevy) [inline app/helpers/smart_proxies_helper.rb]: this should probably be cached to save the many queries, i thought i saw it in the pr but it was gone - @xprazak2 ?

File: app/services/proxy_status/dhcp.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
+module ProxyStatus
+  class DHCP < Base
+
+    def subnets
+      return @subnets if @subnets
+
+      fetch_proxy_data do
+        @subnets = api.subnets.map do |subnet|
+          records = api.subnet(subnet['network'])
```
Comment (ohadlevy) [inline app/services/proxy_status/dhcp.rb:9]: this will be potentially slow with many subnets, the alternative is to create another controller serving subnets (as similar to the dhcp proxy plugin) what do you think @xprazak2 ?

File: app/helpers/smart_proxies_helper.rb (hunk 13 above)
Comment (xprazak2) [inline app/helpers/smart_proxies_helper.rb]: It was [there](https://github.com/xprazak2/foreman/commit/ab8b4a70969cf460ffd57b50a6373c7cebbe9964#diff-5b769784fd39cddb57cf503cd6f2392eR88) but somehow the earlier version got merged, creating a [fix](https://github.com/ohadlevy/foreman/compare/sp-dhcp...xprazak2:host-query?expand=1).

File: app/services/proxy_status/dhcp.rb (hunk 14 above)
Comment (xprazak2) [inline app/services/proxy_status/dhcp.rb:9]: It is a possibility, seems to me a little less intrusive than creating a new action directly in the DHCP module but we will be adding stuff into proxy's API nevertheless, so I think @witlessbird can tell us if it is a good idea.

Comment (ohadlevy) [inline app/services/proxy_status/dhcp.rb:9]: @xprazak2 in the code above you are hitting two different api endpoints, first we fetch / (return a list of networks/subnets) and another one /network_id for each netwrok returned from the first query.

if we want to ensure this works with a large setup, we could do something puppet ca controller, which then can fetch / display each parameter at a time. we can later add a reservation controller, which can show each reservation / edit / destroy etc (similar to https://github.com/theforeman/foreman_dhcp_browser/blob/master/app/controllers/dhcp_controller.rb)

I don't think we need to change anything on the smart proxy api?

Comment (xprazak2) [inline app/services/proxy_status/dhcp.rb:9]: Oh, I misunderstood. You are right, we do not need to change smart proxy.

**Comment by xprazak2** [comment]
There is a new [PR](https://github.com/ohadlevy/foreman/pull/8) against this branch that should take care of possible performance issues for large nuber of subnets and hosts.

File: app/views/smart_proxies/plugins/_dhcp_reservations.html.erb [hunk 15]
```diff
@@ -0,0 +1,18 @@
+<table id="<%= subnet.network.parameterize %>_reservations" class='<%= table_css_classes %> table-fixed' data-table="inline">
+  <thead>
+    <tr>
+      <th class="col-md-4"><%= _("Hostname") %></th>
+      <th class="col-md-2"><%= _("IP Address") %></th>
+      <th class="col-md-2"><%= _("MAC") %></th>
```
Comment (lzap) [inline app/views/smart_proxies/plugins/_dhcp_reservations.html.erb:6]: That does not sum to 12, isn't that an issue?

**Comment by lzap** [comment]
So where we are now with this patch? I would love to see this in 1.11. If Ondřej is taking over this, I'd recommend to move to a different PR completely, nested discussion is not good.

I have no problems with two levels of tabs, Puppet already uses this and it was merged, I'd follow that pattern and we can get UX feedback for the next release. An alternative approach could be a drop down instead of 2nd level of tabs, but again since we already have this in develop...

File: app/views/smart_proxies/plugins/_dhcp_reservations.html.erb (hunk 15 above)
Comment (xprazak2) [inline app/views/smart_proxies/plugins/_dhcp_reservations.html.erb:6]: It displays ok for me: 

but you are probably right, I will make it add up to 12 just to be on a safe side.

**Comment by ohadlevy** [comment]
closed in favor of #3166
